Printing formati u stringovima
String formating nam omogucava ubacivanje nekih stvari u string sto je bolje nego da pokusavamo da ih nadovezujemo jedan na drugi. Brza provera brzine je u sledevem primeru 

    igrac="tomson"
    poeni=33

    'Prosle veceri' +'igrac'+'je postigao' +str(poeni) +'poena'.   #concatenation

    f'Prose veceri {igrac} je postigao {poeni} poena.'             # string format
Postojue tri nacina za string formatiranje:

1. Najstariji metod koriscenje placeholdera koriscenjem <code>%</code> karaktera
2. Unapredjena tehnika je kosricenje <code>.format()</code> 
3. Najnovija metoda je koriscenje f-stringa

## Formatiranje sa placeholderom

Mozemo koristit <code>%s</code> za ubacivanje elemenata u string prlikom printovanja. Karakter `%` nam omogucava ovo svojstvo.

In [1]:
print("I'm going to inject %s here." %'something')

I'm going to inject something here.


Mozemo proslediti vise stvari placeholderu.

In [2]:
print("I'm going to inject %s text here, and %s text here." %('some','more'))

I'm going to inject some text here, and more text here.


MOzemo proslediti i ime promenljive:

In [3]:
x, y = 'some', 'more'
print("I'm going to inject %s text here, and %s text here."%(x,y))

I'm going to inject some text here, and more text here.


### Konverzacijke metode prilikom formatiranja.
Moze se priometiti da dve metode <code>%s</code> i <code>%r</code> konvertuju bilo koji pajton objekat u string koristeci dve potpuno razdvojena metode: `str()` i `repr()`

In [4]:
print('He said his name was %s.' %'Fred')
print('He said his name was %r.' %'Fred')

He said his name was Fred.
He said his name was 'Fred'.


Jos jedan primer, `\t` je ubacivanje tab(razmaka) u string.

In [5]:
print('I once caught a fish %s.' %'this \tbig')
print('I once caught a fish %r.' %'this \tbig')

I once caught a fish this 	big.
I once caught a fish 'this \tbig'.


`%s` konvertuje bilo sta iz pajtona u string, ukljuvcujucu i integer i float borjeve. Dok `%d`konverutje prvo u integer bez zaokruzivanja. Npr:

In [7]:
print('I wrote %s programs today.' %3.75)
print('I wrote %d programs today.' %3.75)   

I wrote 3.75 programs today.
I wrote 3 programs today.


### Zaokruzivanje brojeva
Brojevi sa pokretnim zarezom se mogu zaokruzivati koristeci metodiu <code>%5.2f</code>.Ovde <code>5</code> predstvala minimalan broj ciara levo od zareza,gde se mesta popunjavaju praznim mestima ako imamo manje cifara a <code>.2f</code> predstavlja broj cifara desno od zareza. Npr:

In [10]:
print('Float brojevi: %5.2f' %(13.144))

Float brojevi: 13.14


In [11]:
print('Float brojevi: %1.0f' %(13.144))

Float brojevi: 13


In [12]:
print('Float brojevi: %1.5f' %(13.144))

Float brojevi: 13.14400


In [13]:
print('Float brojevi: %10.2f' %(13.144))

Float brojevi:      13.14


In [14]:
print('Float brojevi: %25.2f' %(13.144))

Float brojevi:                     13.14


### Visestruko formatiranje

In [15]:
print('Prvi: %s, Drugi: %5.2f, Treci: %r' %('Hello!',3.1415,'bye!'))

Prvi: Hello!, Drugi:  3.14, Treci: 'bye!'


## Formatiranje sa  `.format()` metodom
 sintaksa `.format()` method je:

    'String here {} then also {}'.format('something1','something2')
    
npr:

In [16]:
print('Ovo je string sa {}'.format('UBACIVANJEM'))

Ovo je string sa UBACIVANJEM


### Metod .format() ima nekoliko prednosti u odnosu na  %s placeholder mrtod:

#### 1.Uneseni objekti se mogu pozivati pozicionim indeksima:

In [18]:
print('{2} {1} {0}'.format('lisica','braon','brza'))

brza braon lisica


#### 2. Ubacivanim elementima se mogu dodeliti kljucne reci: 

In [19]:
print('Prvi objekat: {a}, Drugi objekat: {b}, Treci objekat: {c}'.format(a=1,b='dva',c=12.3))

Prvi objekat: 1, Drugi objekat: dva, Treci objekat: 12.3


#### 3. Unoseni elementi se mogu ponovo unositi, bez duplikata  u metodi:

In [23]:
print('Sacuvani %s je zaradjen %s dinar.' %('dinar','dinar'))
# vs.
print('Sacuvani {p} je zaradjen {p} dinar.'.format(p='dinar'))

Sacuvani dinar je zaradjen dinar dinar.
Sacuvani dinar je zaradjen dinar dinar.


### Poravnjanje, pedingandpreciznost sa `.format()`
Sa viticastim zagradama mozemo precizirati duzinu polja, levo/desno poravnjanje i jos toga


In [24]:
print('{0:8} | {1:9}'.format('Voce', 'Kolicina'))
print('{0:8} | {1:9}'.format('jabuke', 3.))
print('{0:8} | {1:9}'.format('Narandze', 10))

Voce     | Kolicina 
jabuke   |       3.0
Narandze |        10


Po difoltu `.format()` poravnjanje teksta je levo, brojevi desno. Mozemo prosediti opciono `<`,`^`, ili `>` da podesimo levo, desno ili centar poravnjanja

In [25]:
print('{0:<8} | {1:^8} | {2:>8}'.format('Levo','Centar','Desno'))
print('{0:<8} | {1:^8} | {2:>8}'.format(11,22,33))

Levo     |  Centar  |    Desno
11       |    22    |       33


Mozemo koristiti neki karakter sa peding karakterom

In [27]:
print('{0:=<8} | {1:-^8} | {2:.>8}'.format('levo','Centar','desno'))
print('{0:=<8} | {1:-^8} | {2:.>8}'.format(11,22,33))

levo==== | -Centar- | ...desno
11====== | ---22--- | ......33


Velicina polja i float preciznos se moze odraditi slicno kao i sa placeholdeima.

In [29]:
print('Ovo je 10  mesta dvo-cifatnog broja:%10.2f' %13.579)
print('Ovo je 10  mesta dvo-cifatnog broja:{0:10.2f}'.format(13.579))

Ovo je 10 praznih mesta dvo-cifatnog broja:     13.58
Ovo je 10 praznih mesta dvo-cifatnog broja:     13.58


Primetimo da ima 5 pranih mesta pre znaka <code>:</code>, i 5 pet mesta koje je zauzeto sa 13.58 za ukopno 10 karaktera

## Formatiranje stringa sa metodom:  f-strings

In [30]:
name = 'Marko'

print(f"On je rekao da se zove {name}.")

On je rekao da se zove Marko.


Dodajmo `!r` da bismo predstavili u obliku stringa:

In [34]:
print(f"On je rekao da se zove {name!r}.")

On je rekao da se zove 'Marko'.


#### Float formatiranje broja `"result: {value:{width}.{precision}}"`

Tamo gde smo `.format()` metodu radili smo ovako `{value:10.4f}`, sa f-strings to radimo ovako `{value:{10}.{6}}`


In [36]:
broj= 23.45678
print("Deseto-karakterni, cetvorocifarni decimalni broj je:{0:10.4f}".format(broj))
print(f"Deseto-karakterni, cetvorocifarni decimalni broj je:{broj:{10}.{6}}")

Deseto-karakterni, cetvorocifarni decimalni broj je:   23.4568
Deseto-karakterni, cetvorocifarni decimalni broj je:   23.4568


Primetimo da je sa f-string, *preciznost* sredjivana sa totalnim brojem cifara, ne samo sa onim koji prate decimale. F-string ne dodaje cifre koje su potrene

In [38]:
num = 23.45
print("Deseto-karakterni, cetvorocifarni decimalni broj je:{0:10.4f}".format(num))
print(f"Deseto-karakterni, cetvorocifarni decimalni broj je:{num:{10}.{6}}")

Deseto-karakterni, cetvorocifarni decimalni broj je:   23.4500
Deseto-karakterni, cetvorocifarni decimalni broj je:     23.45


Ako je to potrebno uvek se mozemo vraiti na `.format()`

In [41]:
broj = 23.45
print("Deseto-karakterni, cetvorocifarni decimalni broj je:{0:10.4f}".format(broj))
print(f"Deseto-karakterni, cetvorocifarni decimalni broj je:{broj:10.4f}")

Deseto-karakterni, cetvorocifarni decimalni broj je:   23.4500
Deseto-karakterni, cetvorocifarni decimalni broj je:   23.4500
